File: Python/Binary Trees/L_constStrFromBT.py
def tree2str(root) -> str:
        res = []
        def inOrd(root):
            if not root: return
            res.append("(")
            res.append(str(root.val))
            if not root.left and root.right: res.append("()")
            inOrd(root.left)
            inOrd(root.right)
            res.append(")")
        inOrd(root)
        return "".join(res)[1:-1]


def tree2str(root) -> str:
    def dfs(node):
        if not node.left and not node.right:
            return str(node.val)
        curr = str(node.val)
        if node.left:
            curr += '('+ dfs(node.left) +')'
        if node.right:
            curr += '()' if not node.left else ""
            curr += '('+ dfs(node.right) +')'
        return curr
    return dfs(root)

File: cpp/Binary Tree/L_constStrFromBT.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

string tree2str(TreeNode *root)
{
    string str = "";
    check(root, str);
    return str;
}
void check(TreeNode *root, string &str)
{
    if (root == NULL)
    {
        return;
    }
    str += to_string(root->val);
    if (root->left || root->right)
    {
        str += '(';
        check(root->left, str);
        str += ')';
    }
    if (root->right)
    {
        str += '(';
        check(root->right, str);
        str += ')';
    }
}